x-pack/plugins/beats/server/routes/api/register_assign_tags_to_beats_route.js

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Joi from 'joi';
+import {
+  get,
+  flatten,
+  uniq
+} from 'lodash';
+import { INDEX_NAMES } from '../../../common/constants';
+import { callWithRequestFactory } from '../../lib/client';
+import { wrapEsError } from '../../lib/error_wrappers';
+
+async function getDocs(callWithRequest, ids) {
+  const params = {
+    index: INDEX_NAMES.BEATS,
+    type: '_doc',
+    body: { ids },
+    _source: false
+  };
+
+  const response = await callWithRequest('mget', params);
+  return get(response, 'docs', []);
+}
+
+function getBeats(callWithRequest, beatIds) {
+  const ids = beatIds.map(beatId => `beat:${beatId}`);
+  return getDocs(callWithRequest, ids);
+}
+
+function getTags(callWithRequest, tags) {
+  const ids = tags.map(tag => `tag:${tag}`);
+  return getDocs(callWithRequest, ids);
+}
+
+async function findNonExistentItems(callWithRequest, items, getFn) {
+  const itemsFromEs = await getFn.call(null, callWithRequest, items);
+  return itemsFromEs.reduce((nonExistentItems, itemFromEs, idx) => {
+    if (!itemFromEs.found) {
+      nonExistentItems.push(items[idx]);
+    }
+    return nonExistentItems;
+  }, []);
+}
+
+function findNonExistentBeatIds(callWithRequest, beatIds) {
+  return findNonExistentItems(callWithRequest, beatIds, getBeats);
+}
+
+function findNonExistentTags(callWithRequest, tags) {
+  return findNonExistentItems(callWithRequest, tags, getTags);
+}
+
+async function persistAssignments(callWithRequest, assignments) {
+  const body = flatten(assignments.map(assignment => {
```

Comment:
Could we prefer `array.prototype.flatten` here over the `lodash` function?

Reply:
Link to this function? I found https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatten but that looks like its experimental and not yet supported in Node.js.

I also tried:

```
$ node
> const a = [ [ 1, 2 ], [ 3, 4 ] ]
undefined
> a
[ [ 1, 2 ], [ 3, 4 ] ]
> a.flat()
TypeError: a.flat is not a function
> a.flatten()
TypeError: a.flatten is not a function
>
```

Reply:
Sorry, sorry, I think it's `array.prototype.flat`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatten

And we're probably safe to use lodash as I just realized it's experimental.

Reply:
I only thought to try to avoid lodash here because I was able to circumvent it recently using the spread operator.

Reply:
No worries. If you can think of a way to avoid lodash here while keeping the code readable, let me know and I'm happy to implement it. For now, are you okay with me keeping the lodash usage here?

Reply:
Oh yes, unless there's an alternative I'm fine with `lodash` here.